File: app/src/main/java/com/fenghuang/caipiaobao/widget/ijkplayer/controller/player/widget/PIPManager.java
package com.fenghuang.caipiaobao.widget.ijkplayer.controller.player.widget;

import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;

import com.fenghuang.caipiaobao.app.CaiPiaoBaoApplication;
import com.fenghuang.caipiaobao.widget.ijkplayer.controller.player.DanmukuVideoView;

/**
 * @ Author  QinTian
 * @ Date  2019/11/19- 14:55
 * @ Describe
 */
public class PIPManager {
    private static PIPManager instance;
    private DanmukuVideoView mVideoView;
    private FloatView floatView;
    private FloatController mFloatController;
    private boolean isShowing;
    //    private KeyReceiver mKeyReceiver;
    private int mPlayingPosition = -1;
    private Class mActClass;
//    private MyVideoListener mMyVideoListener = new MyVideoListener() {
//        @Override
//        public void onComplete() {
//            super.onComplete();
//            reset();
//        }
//    };


    private PIPManager() {
        mVideoView = new DanmukuVideoView(CaiPiaoBaoApplication.Companion.getInstance());
//        mVideoView.setVideoListener(mMyVideoListener);
//        mKeyReceiver = new KeyReceiver();
        mFloatController = new FloatController(CaiPiaoBaoApplication.Companion.getInstance());
//        IntentFilter homeFilter = new IntentFilter(Intent.ACTION_CLOSE_SYSTEM_DIALOGS);
//        MyApplication.getInstance().registerReceiver(mKeyReceiver, homeFilter);
        floatView = new FloatView(CaiPiaoBaoApplication.Companion.getInstance(), 0, 0);
    }

    public static PIPManager getInstance() {
        if (instance == null) {
            synchronized (PIPManager.class) {
                if (instance == null) {
                    instance = new PIPManager();
                }
            }
        }
        return instance;
    }

    public DanmukuVideoView getVideoView() {
        return mVideoView;
    }

    public void startFloatWindow() {
        if (isShowing) return;
        removePlayerFormParent();
        mFloatController.setPlayState(mVideoView.getCurrentPlayState());
        mFloatController.setPlayerState(mVideoView.getCurrentPlayerState());
        mVideoView.setVideoController(mFloatController);
        floatView.addView(mVideoView);
        floatView.addToWindow();
        isShowing = true;
    }

    public void stopFloatWindow() {
        if (!isShowing) return;
        floatView.removeFromWindow();
        removePlayerFormParent();
        isShowing = false;
    }

    private void removePlayerFormParent() {
        ViewParent parent = mVideoView.getParent();
        if (parent instanceof ViewGroup) {
            ((ViewGroup) parent).removeView(mVideoView);
        }
    }

    public int getPlayingPosition() {
        return mPlayingPosition;
    }

    public void setPlayingPosition(int position) {
        this.mPlayingPosition = position;
    }

    public void pause() {
        if (isShowing) return;
        mVideoView.pause();
    }

    public void resume() {
        if (isShowing) return;
        mVideoView.resume();
    }

    public void reset() {
        if (isShowing) return;
        removePlayerFormParent();
        mVideoView.release();
        mVideoView.setVideoController(null);
        mPlayingPosition = -1;
        mActClass = null;
    }

    public boolean onBackPress() {
        return !isShowing && mVideoView.onBackPressed();
    }

    public boolean isStartFloatWindow() {
        return isShowing;
    }

    /**
     * 显示悬浮窗
     */
    public void setFloatViewVisible() {
        if (isShowing) {
            mVideoView.resume();
            floatView.setVisibility(View.VISIBLE);
        }
    }

    public Class getActClass() {
        return mActClass;
    }

    public void setActClass(Class cls) {
        this.mActClass = cls;
    }

}

File: app/src/main/java/com/fenghuang/caipiaobao/utils/palyer/PIPManager.kt
package com.fenghuang.caipiaobao.utils.palyer

import android.content.Context
import android.view.View
import android.view.ViewGroup

import com.fenghuang.baselib.base.fragment.BaseFragment
import com.fenghuang.caipiaobao.app.CaiPiaoBaoApplication
import com.fenghuang.caipiaobao.utils.palyer.widget.FloatController
import com.fenghuang.caipiaobao.utils.palyer.widget.FloatView
import com.fenghuang.caipiaobao.widget.ijkplayer.controller.widget.DefinitionVideoView

/**
 * 悬浮播放
 * Created by Devlin_n on 2018/3/30.
 */

class PIPManager {
    val mVideoView: DefinitionVideoView = DefinitionVideoView(CaiPiaoBaoApplication.getInstance())
    private val mFloatView: FloatView = FloatView(CaiPiaoBaoApplication.getInstance(), 0, 0)
    private val mFloatController: FloatController = FloatController(CaiPiaoBaoApplication.getInstance())
    var isStartFloatWindow: Boolean = false
        private set
    var playingPosition = -1
    var actClass: Class<*>? = null
    var mFragment: BaseFragment? = null
    var mContext: Context? = null


    fun startFloatWindow() {
        if (isStartFloatWindow) return
        removePlayerFormParent()
        mFloatController.setPlayState(mVideoView.currentPlayState)
        mFloatController.setPlayerState(mVideoView.currentPlayerState)
        mVideoView.setVideoController(mFloatController)
        mFloatView.addView(mVideoView)
        mFloatView.addToWindow()
        isStartFloatWindow = true
    }

    fun stopFloatWindow() {
        if (!isStartFloatWindow) return
        mFloatView.removeFromWindow()
        removePlayerFormParent()
        isStartFloatWindow = false
    }

    private fun removePlayerFormParent() {
        val parent = mVideoView.parent
        if (parent is ViewGroup) {
            parent.removeView(mVideoView)
        }
    }

    fun pause() {
        if (isStartFloatWindow) return
        mVideoView.pause()
    }

    fun resume() {
        if (isStartFloatWindow) return
        mVideoView.resume()
    }

    fun reset() {
        if (isStartFloatWindow) return
        removePlayerFormParent()
        mVideoView.setVideoController(null)
        mVideoView.release()
        playingPosition = -1
        actClass = null
        mFragment = null
    }

    fun onBackPress(): Boolean {
        return !isStartFloatWindow && mVideoView.onBackPressed()
    }

    /**
     * 显示悬浮窗
     */
    fun setFloatViewVisible() {
        if (isStartFloatWindow) {
            mVideoView.resume()
            mFloatView.visibility = View.VISIBLE
        }
    }

    fun setContext(context: Context) {
        mContext = context
    }

    companion object {

        private var instance: PIPManager? = null

        fun getInstance(): PIPManager {
            if (instance == null) {
                synchronized(PIPManager::class.java) {
                    if (instance == null) {
                        instance = PIPManager()
                    }
                }
            }
            return instance!!
        }
    }

}
